compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUSelectUKernels.cpp

```diff
@@ -158,39 +169,200 @@ static ArrayAttr lookUpExecutableObjects(Operation *op,
+    llvm::GenericValue v;
+    v.IntVal = APInt(32, val);
+    return v;
+  };
+  SmallVector<llvm::GenericValue> args{
+      constI32(mma.getIntrinsicsM()), constI32(mma.getSubgroupsM()),
+      constI32(mma.getIntrinsicsN()), constI32(mma.getSubgroupsN()),
+      constI32(mma.getIntrinsicsK())};
+  if (func->arg_size() != args.size()) {
+    op.emitWarning(
+        llvm::formatv("Bitcode function {} takes {} arguments. Expected {}.",
+                      queryFuncName, func->arg_size(), args.size()));
+    return {};
+  }
+  llvm::GenericValue interpreterResult = interpreter->runFunction(func, args);
+  if (interpreter->hasError()) {
+    op.emitWarning(llvm::formatv("Error while interpreting bitcode: {}.",
+                                 interpreter->getErrorMessage()));
+    return {};
+  }
+  int sharedMemoryBytes = interpreterResult.IntVal.getSExtValue();
+
+  // Reject a ukernel that would consume too much shared memory, which we need
+  // to save for other purposes. This threshold can always be adjusted but we
+  // default to a low threshold to get an early signal.
+  int maxSharedMemoryBytes = getSharedMemoryBytes(gpuTarget) / 4;
+  if (sharedMemoryBytes > maxSharedMemoryBytes) {
+    op.emitWarning(llvm::formatv("The shared memory size {} required by the "
+                                 "ukernel exceeds the maximum allowed size {}.",
+                                 sharedMemoryBytes, maxSharedMemoryBytes));
+    return {};
+  }
+  return sharedMemoryBytes;
+}
+
+// Returns the shared memory size required by the multi_mma ukernel.
+// On failure, an op warning is emitted and {} is returned.
+// Uses a static cache to avoid calling expensivelyEvaluateSharedMemoryBytes
+// more than once per DataTiledMMAAttr value.
+static std::optional<int>
+getSharedMemoryBytes(IREE::GPU::MultiMmaOp op,
+                     IREE::GPU::UKernelConfigAttr ukernelConfig,
+                     IREE::HAL::ExecutableObjectAttr bitcodeObject,
+                     IREE::GPU::TargetAttr gpuTarget) {
+  auto mma = dyn_cast<IREE::GPU::DataTiledMMAAttr>(op.getKind());
+
+  // We use the stringification of the attributes, rather than the
+  // attributes themselves, as the key, to ensure it's self-contained and does
+  // not contain pointers to other objects, such as a `MLIRContext*`, which
+  // could go dangling.
+  std::string key = llvm::formatv("mma = {}, gpuTarget = {}", mma, gpuTarget);
+
+  // The cache and the mutex guarding it.
+  static llvm::StringMap<std::optional<int>> cachedSharedMemoryBytess;
+  static std::mutex mutex;
+
+  {
+    // Critical section. Query the cache.
+    std::lock_guard<std::mutex> lock(mutex);
+    if (cachedSharedMemoryBytess.contains(key)) {
+      return cachedSharedMemoryBytess.at(key);
+    }
+  }
+
+  // Not found in cache. Perform the expensive evaluation.
+  std::optional<int> sharedMemoryBytes = expensivelyEvaluateSharedMemoryBytes(
+      op, ukernelConfig, bitcodeObject, gpuTarget);
```

Comment:
With this being outside of a critical section, I think it's possible for you to run the interpreter multiple times in parallel for the same input. Shouldn't we put the query & calculation & insertion into a single critical section to avoid the interpreter overhead?

Reply:
I see the point when multiple threads are concurrently doing the same evaluation.

But there's an opposite scenario when multiple threads are concurrently doing distinct evaluations. Currently they can run in parallel, but with a single critical section, they are all waiting on the one thread doing the evaluation. Unlike the case where multiple threads are doing the same evaluation, here there is no benefit in waiting.

While real user workloads are likely to have few DataTiledMMAAttr values (so the case you're thinking about is the most important one), there is a plausible use case for benchmarking/tuning purposes where a single program contains many functions, each with its own separate DataTiledMMAAttr. We shouldn't serialize it.

Reply:
A smarter scheme would be to have the first first thread insert a placeholder entry (for its key) for not yet calculated sizes and then have the other threads wait on that. But that would require a condition variable, and is probably ah overkill until we learn otherwise.

Reply:
I'm still trying to think of a solution that addresses both opposite concerns, without introducing too much complexity.

Reply:
I implemented something like this in the past: https://github.com/GPUOpen-Drivers/llpc/blob/6aecd3da132c62e9b38c1596f62865b8178a73d6/llpc/tool/llpcShaderCache.cpp#L476-L488

Reply:
How about this additional commit I just pushed to this PR (so we can always drop it if you don't like it?). No CondVars, but separate mutexes per entry in addition to the big one guarding the hash map.
https://github.com/iree-org/iree/pull/19786/commits/14d8fd5688f04e04e7fa8c92e00cf3e4654bdbfa